File: src/scrape_calculations.py
import os
import json
import argparse
import asyncio
import aiohttp
import aiofiles
import boto3
import logging
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Optional
from botocore.exceptions import ClientError
from bs4 import BeautifulSoup
import re

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

DEFAULT_AWS_REGION = "us-east-2"
BASE_URL = "https://ratings.fide.com/a_indv_calculations.php"
TIME_CONTROLS = {"standard": "0", "rapid": "1", "blitz": "2"}

class PlayerCalculationScraper:
    def __init__(self, s3_bucket, aws_region=DEFAULT_AWS_REGION):
        self.s3_bucket = s3_bucket
        self.s3_client = boto3.client('s3', region_name=aws_region)
        self.local_temp_dir = "/tmp/chess_data"
        
        # High concurrency for this intensive step, but respect server limits
        self.max_concurrent_requests = 30
        self.semaphore = asyncio.Semaphore(self.max_concurrent_requests)
        
        logger.info(f"Initialized calculation scraper - S3 bucket: {s3_bucket}, Region: {aws_region}")
        logger.info(f"Max concurrent requests: {self.max_concurrent_requests}")

    async def scrape_calculations_for_month(self, month_str):
        """Scrape calculation data for all active players for a specific month"""
        try:
            year, month = map(int, month_str.split("-"))
            logger.info(f"Starting calculation scraping for {year}-{month:02d}")
            
            # Setup local temp directory
            os.makedirs(self.local_temp_dir, exist_ok=True)
            
            # Determine which time controls to process based on year
            time_controls_to_process = ["standard"]
            if year >= 2012:
                time_controls_to_process.extend(["rapid", "blitz"])
            
            logger.info(f"Processing time controls: {time_controls_to_process}")
            
            # Configure async session with higher limits for this intensive step
            connector = aiohttp.TCPConnector(
                limit=200,          # Total connection pool size
                limit_per_host=50,  # Connections per host (FIDE server)
                ttl_dns_cache=300,  # DNS cache TTL
                use_dns_cache=True
            )
            timeout = aiohttp.ClientTimeout(total=60)
            
            async with aiohttp.ClientSession(connector=connector, timeout=timeout) as session:
                # Process each time control in parallel
                tasks = [
                    self._process_time_control(session, tc, year, month)
                    for tc in time_controls_to_process
                ]
                
                logger.info(f"Processing {len(time_controls_to_process)} time controls in parallel")
                
                # Execute all time control tasks concurrently
                results = await asyncio.gather(*tasks, return_exceptions=True)
                
                # Process results and log summary
                successful = []
                failed = []
                
                for i, result in enumerate(results):
                    tc = time_controls_to_process[i]
                    if isinstance(result, tuple) and result[0] == "success":
                        successful.append((tc, result[1], result[2]))  # (tc, processed, successful)
                    elif result == "skipped":
                        logger.info(f"No active players found for {tc}")
                    else:
                        failed.append((tc, str(result) if isinstance(result, Exception) else result))
                
                total_processed = sum(processed for _, processed, _ in successful)
                total_successful = sum(successful_count for _, _, successful_count in successful)
                
                logger.info(f"Results - Time controls: {len(successful)} successful, {len(failed)} failed")
                logger.info(f"Total calculations: {total_processed} processed, {total_successful} successful")
                
                # Log failures for debugging
                if failed:
                    logger.error(f"Failed time controls: {[f[0] for f in failed]}")
                    for tc, error in failed:
                        logger.error(f"  {tc}: {error}")
                
                # Upload completion marker
                await self._upload_completion_marker(month_str, len(successful), len(failed), successful)
                
                return True
                
        except Exception as e:
            logger.error(f"Error in calculation scraping: {str(e)}", exc_info=True)
            raise

    async def _process_time_control(self, session: aiohttp.ClientSession, time_control: str, year: int, month: int):
        """Process all players for a specific time control"""
        month_str = f"{month:02d}"
        
        try:
            # Download active player list from S3
            player_ids = await self._download_active_players(time_control, year, month)
            
            if not player_ids:
                logger.info(f"No active players found for {time_control} {year}-{month_str}")
                return "skipped"
            
            logger.info(f"Processing {len(player_ids)} players for {time_control}")
            
            # Process players in optimized batches
            batch_size = 100  # Process 100 players at a time
            total_batches = (len(player_ids) + batch_size - 1) // batch_size
            total_processed = 0
            total_successful = 0
            
            for i in range(0, len(player_ids), batch_size):
                batch = player_ids[i:i + batch_size]
                batch_num = i // batch_size + 1
                
                logger.info(f"{time_control}: Processing batch {batch_num}/{total_batches} ({len(batch)} players)")
                
                # Create tasks for this batch
                tasks = [
                    self._process_single_player(session, player_id, time_control, year, month)
                    for player_id in batch
                ]
                
                # Execute batch concurrently
                results = await asyncio.gather(*tasks, return_exceptions=True)
                
                # Count results
                batch_successful = sum(1 for result in results if result is True)
                batch_processed = len(results)
                
                total_processed += batch_processed
                total_successful += batch_successful
                
                logger.info(f"{time_control}: Batch {batch_num} completed - "
                          f"{batch_successful}/{batch_processed} successful, "
                          f"Total: {total_successful}/{total_processed}")
                
                # Brief pause between batches to be respectful to the server
                if batch_num < total_batches:
                    await asyncio.sleep(0.5)
            
            logger.info(f"{time_control}: Completed - {total_successful}/{total_processed} players processed successfully")
            
            return ("success", total_processed, total_successful)
            
        except Exception as e:
            logger.error(f"Error processing time control {time_control}: {str(e)}")
            return e

    async def _download_active_players(self, time_control: str, year: int, month: int) -> List[str]:
        """Download active player list for a time control from S3"""
        month_str = f"{month:02d}"
        s3_key = f"persistent/active_players/{year}-{month_str}_{time_control}.txt"
        
        try:
            loop = asyncio.get_event_loop()
            response = await loop.run_in_executor(
                None,
                lambda: self.s3_client.get_object(Bucket=self.s3_bucket, Key=s3_key)
            )
            
            content = response['Body'].read().decode('utf-8')
            player_ids = [line.strip() for line in content.splitlines() if line.strip()]
            
            return player_ids
            
        except ClientError as e:
            if e.response['Error']['Code'] == 'NoSuchKey':
                logger.info(f"No active players file found for {time_control} {year}-{month_str}")
                return []
            else:
                logger.error(f"Error downloading active players for {time_control}: {str(e)}")
                return []
        except Exception as e:
            logger.error(f"Unexpected error downloading active players for {time_control}: {str(e)}")
            return []

    async def _process_single_player(self, session: aiohttp.ClientSession, player_id: str, time_control: str, year: int, month: int) -> bool:
        """Process calculation data for a single player"""
        month_str = f"{month:02d}"
        
        # Check if already processed
        s3_key = f"persistent/calculations/{year}-{month_str}/{time_control}/{player_id}.json"
        if await self._check_s3_file_exists_async(s3_key):
            return True  # Already processed
        
        tc_code = TIME_CONTROLS[time_control]
        period_formatted = f"{year}-{month_str}-01"
        url = f"{BASE_URL}?id_number={player_id}&rating_period={period_formatted}&t={tc_code}"
        
        max_retries = 3
        base_delay = 1
        
        async with self.semaphore:
            for attempt in range(max_retries):
                try:
                    async with session.get(url, timeout=30) as response:
                        response.raise_for_status()
                        html_content = await response.text()

                        # Check if there's actual calculation data
                        if ("No calculations available" in html_content or 
                            "No games" in html_content or 
                            len(html_content.strip()) < 100):
                            # No calculation data - this is normal for many players
                            return True

                        # Extract structured data from the HTML
                        calculation_data = self._extract_calculation_data(html_content)

                        if not calculation_data.get("tournaments"):
                            # No tournaments found in the calculation
                            return True

                        # Save to local file first
                        local_file = await self._save_calculation_locally(calculation_data, player_id, time_control, year, month)
                        
                        # Upload to S3
                        await self._upload_to_s3_async(local_file, s3_key)
                        
                        # Cleanup local file
                        os.remove(local_file)
                        
                        return True

                except (aiohttp.ClientError, asyncio.TimeoutError) as e:
                    delay = base_delay * (2 ** attempt)
                    if attempt < max_retries - 1:
                        await asyncio.sleep(delay)
                    else:
                        logger.warning(f"Failed to process player {player_id} after {max_retries} attempts: {e}")
                        return False
                except Exception as e:
                    logger.error(f"Unexpected error processing player {player_id}: {str(e)}")
                    return False

        return False

    def _extract_calculation_data(self, html_content: str) -> Dict:
        """Extract player calculation data from HTML content"""
        soup = BeautifulSoup(html_content, "html.parser")
        result_data = {
            "tournaments": [],
        }

        # First find all tournament headers
        tournament_headers = soup.find_all("div", class_="rtng_line01")
        tournament_ids = []
        tournament_names = []

        for header in tournament_headers:
            # Look for tournament links and extract IDs
            tournament_link = header.find("a")
            if tournament_link:
                href = tournament_link.get("href", "")
                tournament_id_match = re.search(r"event=(\d+)", href)
                if tournament_id_match:
                    tournament_id = tournament_id_match.group(1)
                    tournament_name = tournament_link.text.strip()
                    tournament_ids.append(tournament_id)
                    tournament_names.append(tournament_name)

        # Find all calculation tables
        calc_tables = soup.find_all("table", class_="calc_table")

        # Process each table and associate with tournament ID
        for i, calc_table in enumerate(calc_tables):
            if i < len(tournament_ids):  # Make sure we have a tournament ID for this table
                tournament_id = tournament_ids[i]

                # Create tournament entry
                current_tournament = {"tournament_id": tournament_id, "games": []}

                # Check if player is unrated in this tournament
                rp_headers = calc_table.find_all("td", string=lambda s: s and "Rp" in s)
                current_tournament["player_is_unrated"] = len(rp_headers) > 0

                # Find all rows with game data
                game_rows = calc_table.find_all("tr", attrs={"bgcolor": "#efefef"})

                for row in game_rows:
                    try:
                        cells = row.find_all("td", class_="list4")
                        if len(cells) < 6:  # Need at least name, rating, federation, result
                            continue

                        game = {}

                        # Extract opponent name
                        opponent_cell = cells[0]
                        game["opponent_name"] = opponent_cell.text.strip()

                        # Extract rating
                        if len(cells) > 3:
                            rating_text = cells[3].text.strip()
                            rating = re.search(r"-?\d+", rating_text)
                            if rating:
                                game["opponent_rating"] = rating.group()

                        # Extract federation
                        if len(cells) > 4:
                            game["federation"] = cells[4].text.strip()

                        # Extract game result
                        if len(cells) > 5:
                            game["result"] = cells[5].text.strip()

                        # Add tournament_id to each game
                        game["tournament_id"] = tournament_id

                        # Add to current tournament's games
                        current_tournament["games"].append(game)

                    except Exception as e:
                        logger.warning(f"Error extracting game data: {e}")

                # Add tournament to results if it has games
                if current_tournament["games"]:
                    result_data["tournaments"].append(current_tournament)

        return result_data

    async def _save_calculation_locally(self, calculation_data: Dict, player_id: str, time_control: str, year: int, month: int) -> str:
        """Save calculation data to local file"""
        month_str = f"{month:02d}"
        local_dir = os.path.join(self.local_temp_dir, "calculations", f"{year}-{month_str}", time_control)
        os.makedirs(local_dir, exist_ok=True)
        
        local_file = os.path.join(local_dir, f"{player_id}.json")
        
        async with aiofiles.open(local_file, "w", encoding="utf-8") as f:
            await f.write(json.dumps(calculation_data, indent=2))
        
        return local_file

    async def _check_s3_file_exists_async(self, s3_key: str) -> bool:
        """Async wrapper for checking S3 file existence"""
        loop = asyncio.get_event_loop()
        try:
            await loop.run_in_executor(
                None, 
                lambda: self.s3_client.head_object(Bucket=self.s3_bucket, Key=s3_key)
            )
            return True
        except ClientError:
            return False

    async def _upload_to_s3_async(self, local_path: str, s3_key: str):
        """Async wrapper for S3 upload"""
        loop = asyncio.get_event_loop()
        try:
            await loop.run_in_executor(
                None, 
                lambda: self.s3_client.upload_file(local_path, self.s3_bucket, s3_key)
            )
        except ClientError as e:
            logger.error(f"Failed to upload to S3: {str(e)}")
            raise

    async def _upload_completion_marker(self, month_str: str, successful: int, failed: int, success_details: List):
        """Upload completion marker with statistics"""
        try:
            completion_data = {
                "month": month_str,
                "timestamp": datetime.utcnow().isoformat() + "Z",
                "statistics": {
                    "time_controls_processed": successful + failed,
                    "time_controls_successful": successful,
                    "time_controls_failed": failed,
                    "details": {tc: {"processed": processed, "successful": success_count} 
                              for tc, processed, success_count in success_details}
                },
                "step": "calculation_scraping",
                "method": "high_concurrency_batch"
            }
            
            local_file = os.path.join(self.local_temp_dir, f"calculation_scraping_completion_{month_str}.json")
            async with aiofiles.open(local_file, "w") as f:
                await f.write(json.dumps(completion_data, indent=2))
            
            s3_key = f"results/{month_str}/calculation_scraping_completion.json"
            await self._upload_to_s3_async(local_file, s3_key)
            
            os.remove(local_file)
            logger.info(f"Uploaded completion marker for {month_str}")
            
        except Exception as e:
            logger.warning(f"Failed to upload completion marker: {str(e)}")

def main():
    parser = argparse.ArgumentParser(description="Scrape FIDE player calculation data.")
    parser.add_argument(
        "--month",
        type=str,
        help="Month for processing in YYYY-MM format",
        required=True,
    )
    parser.add_argument(
        "--s3_bucket",
        type=str,
        help="S3 bucket for data storage",
        default=os.environ.get("S3_BUCKET"),
        required=True,
    )
    parser.add_argument(
        "--aws_region",
        type=str,
        help="AWS region",
        default=os.environ.get("AWS_REGION", DEFAULT_AWS_REGION),
    )

    args = parser.parse_args()
    
    logger.info(f"Starting calculation scraper for {args.month}")
    
    # Validate month format
    try:
        datetime.strptime(args.month, "%Y-%m")
    except ValueError:
        logger.error("Month must be in YYYY-MM format")
        return 1
    
    # Initialize scraper
    scraper = PlayerCalculationScraper(
        s3_bucket=args.s3_bucket,
        aws_region=args.aws_region
    )
    
    # Run scraping
    try:
        success = asyncio.run(scraper.scrape_calculations_for_month(args.month))
        
        if success:
            logger.info(f"Calculation scraping completed successfully for {args.month}")
            return 0
        else:
            logger.error("Calculation scraping failed")
            return 1
            
    except Exception as e:
        logger.error(f"Calculation scraping failed with error: {str(e)}", exc_info=True)
        return 1

if __name__ == "__main__":
    exit(main())

File: rust-src/src/scrape_calculations.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use anyhow::{Context, Result};
use aws_sdk_s3::Client as S3Client;
use aws_types::region::Region;
use clap::Parser;
use futures::{stream, StreamExt};
use regex::Regex;
use reqwest::Client;
use scraper::{Html, Selector};
use serde::{Deserialize, Serialize};
use serde_json;
use tokio::fs;
use tokio::fs::OpenOptions;
use tokio::io::AsyncWriteExt;
use tokio::sync::Semaphore;
use tokio::time::sleep;
use tracing::{error, info, warn};

#[derive(Parser)]
#[command(name = "calculation-scraper")]
#[command(about = "Scrape FIDE player calculation data")]
struct Args {
    #[arg(long, help = "Month for processing in YYYY-MM format")]
    month: String,
    
    #[arg(long, help = "S3 bucket for data storage")]
    s3_bucket: String,
    
    #[arg(long, default_value = "us-east-2", help = "AWS region")]
    aws_region: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct Game {
    opponent_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    opponent_rating: Option<String>,
    federation: String,
    result: String,
    tournament_id: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct Tournament {
    tournament_id: String,
    player_is_unrated: bool,
    games: Vec<Game>,
}

#[derive(Debug, Serialize, Deserialize)]
struct CalculationData {
    tournaments: Vec<Tournament>,
}

#[derive(Debug, Serialize, Deserialize)]
struct PlayerCalculationRecord {
    player_id: String,
    calculation_data: CalculationData,
}

struct PlayerCalculationScraper {
    s3_client: S3Client,
    http_client: Client,
    s3_bucket: String,
    local_temp_dir: String,
    semaphore: Arc<Semaphore>,
}

impl PlayerCalculationScraper {
    async fn new(s3_bucket: String, aws_region: String) -> Result<Self> {
        let config = aws_config::from_env()
            .region(Region::new(aws_region.clone()))
            .load()
            .await;

        let s3_client = S3Client::new(&config);

        // Initialize HTTP client with conservative settings
        let http_client = Client::builder()
            .pool_max_idle_per_host(30)
            .pool_idle_timeout(Duration::from_secs(60))
            .timeout(Duration::from_secs(45))
            .user_agent("Mozilla/5.0 (compatible; FIDEScraper/1.0)")
            .build()?;

        let local_temp_dir = "/tmp/chess_data".to_string();
        let max_concurrent_requests = 40; // Aggressive but not excessive
        let semaphore = Arc::new(Semaphore::new(max_concurrent_requests));

        info!("Initialized calculation scraper - S3 bucket: {}, Max concurrent: {}", 
              s3_bucket, max_concurrent_requests);

        Ok(Self {
            s3_client,
            http_client,
            s3_bucket,
            local_temp_dir,
            semaphore,
        })
    }

    async fn file_exists_in_s3(&self, key: &str) -> Result<bool> {
        match self.s3_client.head_object().bucket(&self.s3_bucket).key(key).send().await {
            Ok(_) => Ok(true),
            Err(err) => {
                // Check if it's a "not found" error vs actual error
                if let Some(service_err) = err.as_service_error() {
                    if service_err.is_not_found() {
                        return Ok(false);
                    }
                }
                // For other errors, we should probably fail gracefully and continue
                warn!("Error checking if file exists in S3 ({}): {}", key, err);
                Ok(false)
            }
        }
    }

    async fn scrape_calculations_for_month(&self, month_str: &str) -> Result<()> {
        let parts: Vec<&str> = month_str.split('-').collect();
        if parts.len() != 2 {
            return Err(anyhow::anyhow!("Invalid month format. Expected YYYY-MM"));
        }
        
        let year: u32 = parts[0].parse()?;
        let month: u32 = parts[1].parse()?;

        info!("Starting calculation scraping for {}-{:02}", year, month);

        // Check if completion marker already exists
        let completion_marker_key = format!("results/{}/calculation_scraping_completion.json", month_str);
        if self.file_exists_in_s3(&completion_marker_key).await? {
            info!("Completion marker already exists ({}). Skipping entire month processing.", completion_marker_key);
            return Ok(());
        }

        // Create temp directory
        fs::create_dir_all(&self.local_temp_dir).await?;

        // Determine time controls
        let mut time_controls = vec!["standard"];
        if year >= 2012 {
            time_controls.extend_from_slice(&["rapid", "blitz"]);
        }

        info!("Processing time controls: {:?}", time_controls);

        // Process time controls concurrently
        let tasks: Vec<_> = time_controls
            .into_iter()
            .map(|tc| self.process_time_control(tc, year, month))
            .collect();

        let results = futures::future::join_all(tasks).await;

        // Process results
        let mut successful = Vec::new();
        let mut failed = Vec::new();
        let mut skipped = Vec::new();

        for (i, result) in results.into_iter().enumerate() {
            let tc = ["standard", "rapid", "blitz"][i];
            match result {
                Ok(Some((processed, success_count))) => {
                    successful.push((tc.to_string(), processed, success_count));
                }
                Ok(None) => {
                    skipped.push(tc.to_string());
                }
                Err(e) => {
                    failed.push((tc.to_string(), e.to_string()));
                }
            }
        }

        let total_processed: usize = successful.iter().map(|(_, p, _)| p).sum();
        let total_successful: usize = successful.iter().map(|(_, _, s)| s).sum();

        info!("Results - Time controls: {} successful, {} failed, {} skipped", 
              successful.len(), failed.len(), skipped.len());
        info!("Total calculations: {} processed, {} successful", total_processed, total_successful);

        if !failed.is_empty() {
            error!("Failed time controls: {:?}", failed);
        }
        if !skipped.is_empty() {
            info!("Skipped time controls (already exist): {:?}", skipped);
        }

        // Only upload completion marker if we actually processed something or had failures
        if !successful.is_empty() || !failed.is_empty() {
            self.upload_completion_marker(month_str, &successful, &failed).await?;
        } else {
            info!("All time controls were skipped - completion marker already exists or no work needed");
        }
        Ok(())
    }

    async fn process_time_control(&self, time_control: &str, year: u32, month: u32) -> Result<Option<(usize, usize)>> {
        // Check if output file already exists
        let output_key = format!("persistent/calculations/{}-{:02}/{}.jsonl", year, month, time_control);
        if self.file_exists_in_s3(&output_key).await? {
            info!("{}: Output file already exists ({}). Skipping processing.", time_control, output_key);
            return Ok(None);
        }

        let player_ids = self.download_active_players(time_control, year, month).await?;

        if player_ids.is_empty() {
            info!("No active players found for {} {}-{:02}", time_control, year, month);
            return Ok(Some((0, 0)));
        }

        info!("Processing {} players for {}", player_ids.len(), time_control);

        // Create JSONL file for this time control
        let jsonl_file_path = format!("{}/calculations_{}-{:02}_{}.jsonl", 
                                     self.local_temp_dir, year, month, time_control);
        let mut jsonl_file = OpenOptions::new()
            .create(true)
            .truncate(true)
            .write(true)
            .open(&jsonl_file_path)
            .await?;

        let batch_size = 80; // Larger batches for better performance
        let mut total_processed = 0;
        let mut total_successful = 0;

        for (batch_num, chunk) in player_ids.chunks(batch_size).enumerate() {
            let batch_num = batch_num + 1;
            info!("{}: Processing batch {} ({} players)", time_control, batch_num, chunk.len());

            let results: Vec<Option<PlayerCalculationRecord>> = stream::iter(chunk)
                .map(|player_id| self.process_single_player(player_id, time_control, year, month))
                .buffer_unordered(40) // High concurrency within batch
                .collect::<Vec<Result<Option<PlayerCalculationRecord>>>>()
                .await
                .into_iter()
                .map(|r| r.unwrap_or(None))
                .collect();

            let mut batch_successful = 0;
            for result in results {
                total_processed += 1;
                if let Some(record) = result {
                    // Write to JSONL file
                    let json_line = serde_json::to_string(&record)?;
                    jsonl_file.write_all(json_line.as_bytes()).await?;
                    jsonl_file.write_all(b"\n").await?;
                    batch_successful += 1;
                    total_successful += 1;
                }
            }

            info!("{}: Batch {} - {}/{} successful, Total: {}/{}", 
                  time_control, batch_num, batch_successful, chunk.len(), 
                  total_successful, total_processed);

            // Minimal pause to be respectful
            sleep(Duration::from_millis(200)).await;
        }

        // Close the file
        jsonl_file.flush().await?;
        drop(jsonl_file);

        // Upload the consolidated JSONL file to S3
        if total_successful > 0 {
            self.upload_jsonl_file(&jsonl_file_path, time_control, year, month).await?;
        }

        // Clean up local file
        fs::remove_file(&jsonl_file_path).await?;

        info!("{}: Completed - {}/{} players successful", time_control, total_successful, total_processed);
        Ok(Some((total_processed, total_successful)))
    }

    async fn upload_jsonl_file(&self, local_file_path: &str, time_control: &str, year: u32, month: u32) -> Result<()> {
        let file_content = fs::read(local_file_path).await?;
        let s3_key = format!("persistent/calculations/{}-{:02}/{}.jsonl", year, month, time_control);
        
        self.s3_client
            .put_object()
            .bucket(&self.s3_bucket)
            .key(&s3_key)
            .body(file_content.into())
            .send()
            .await?;

        info!("Uploaded JSONL file: {}", s3_key);
        Ok(())
    }

    async fn download_active_players(&self, time_control: &str, year: u32, month: u32) -> Result<Vec<String>> {
        let s3_key = format!("persistent/active_players/{}-{:02}_{}.txt", year, month, time_control);

        match self.s3_client.get_object().bucket(&self.s3_bucket).key(&s3_key).send().await {
            Ok(response) => {
                let body = response.body.collect().await?;
                let content = String::from_utf8(body.to_vec())?;
                let player_ids: Vec<String> = content
                    .lines()
                    .map(|line| line.trim().to_string())
                    .filter(|line| !line.is_empty())
                    .collect();
                Ok(player_ids)
            }
            Err(_) => {
                warn!("No active players file found for {} {}-{:02}", time_control, year, month);
                Ok(Vec::new())
            }
        }
    }

    async fn process_single_player(&self, player_id: &str, time_control: &str, year: u32, month: u32) -> Result<Option<PlayerCalculationRecord>> {
        let tc_code = match time_control {
            "standard" => "0",
            "rapid" => "1", 
            "blitz" => "2",
            _ => return Err(anyhow::anyhow!("Invalid time control: {}", time_control)),
        };

        let url = format!(
            "https://ratings.fide.com/a_indv_calculations.php?id_number={}&rating_period={}-{:02}-01&t={}",
            player_id, year, month, tc_code
        );

        let _permit = self.semaphore.acquire().await?;

        // Single attempt with good timeout
        match self.http_client.get(&url).send().await {
            Ok(response) if response.status().is_success() => {
                let html_content = response.text().await?;

                // Check for no data conditions
                if html_content.contains("No calculations available") ||
                   html_content.contains("No games") ||
                   html_content.trim().len() < 100 {
                    return Ok(None); // Normal case - no data
                }

                // Extract data
                let calculation_data = self.extract_calculation_data(&html_content)?;
                if !calculation_data.tournaments.is_empty() {
                    Ok(Some(PlayerCalculationRecord {
                        player_id: player_id.to_string(),
                        calculation_data,
                    }))
                } else {
                    Ok(None)
                }
            }
            _ => Ok(None), // Failed but don't retry to maintain speed
        }
    }

    fn extract_calculation_data(&self, html_content: &str) -> Result<CalculationData> {
        let document = Html::parse_document(html_content);
        let mut tournaments = Vec::new();
        let mut tournament_ids = Vec::new();

        // Extract tournament IDs
        let header_selector = Selector::parse("div.rtng_line01").unwrap();
        let link_selector = Selector::parse("a").unwrap();
        let event_regex = Regex::new(r"event=(\d+)").unwrap();

        for header in document.select(&header_selector) {
            if let Some(link) = header.select(&link_selector).next() {
                if let Some(href) = link.value().attr("href") {
                    if let Some(captures) = event_regex.captures(href) {
                        if let Some(tournament_id) = captures.get(1) {
                            tournament_ids.push(tournament_id.as_str().to_string());
                        }
                    }
                }
            }
        }

        // Extract game data
        let table_selector = Selector::parse("table.calc_table").unwrap();
        let row_selector = Selector::parse("tr[bgcolor='#efefef']").unwrap();
        let cell_selector = Selector::parse("td.list4").unwrap();
        let rating_regex = Regex::new(r"-?\d+").unwrap();

        for (i, table) in document.select(&table_selector).enumerate() {
            if i >= tournament_ids.len() { break; }

            let tournament_id = &tournament_ids[i];
            let mut games = Vec::new();

            // Check if unrated
            let player_is_unrated = table.inner_html().contains("Rp");

            for row in table.select(&row_selector) {
                let cells: Vec<_> = row.select(&cell_selector).collect();
                if cells.len() < 6 { continue; }

                let opponent_name = cells[0].text().collect::<String>().trim().to_string();
                let opponent_rating = if cells.len() > 3 {
                    let rating_text = cells[3].text().collect::<String>();
                    rating_regex.find(&rating_text).map(|m| m.as_str().to_string())
                } else { None };
                let federation = if cells.len() > 4 { 
                    cells[4].text().collect::<String>().trim().to_string() 
                } else { String::new() };
                let result = if cells.len() > 5 { 
                    cells[5].text().collect::<String>().trim().to_string() 
                } else { String::new() };

                games.push(Game {
                    opponent_name,
                    opponent_rating,
                    federation,
                    result,
                    tournament_id: tournament_id.clone(),
                });
            }

            if !games.is_empty() {
                tournaments.push(Tournament {
                    tournament_id: tournament_id.clone(),
                    player_is_unrated,
                    games,
                });
            }
        }

        Ok(CalculationData { tournaments })
    }

    async fn upload_completion_marker(&self, month_str: &str, successful: &[(String, usize, usize)], failed: &[(String, String)]) -> Result<()> {
        let mut statistics = HashMap::new();
        statistics.insert("time_controls_processed".to_string(), serde_json::Value::Number((successful.len() + failed.len()).into()));
        statistics.insert("time_controls_successful".to_string(), serde_json::Value::Number(successful.len().into()));
        statistics.insert("time_controls_failed".to_string(), serde_json::Value::Number(failed.len().into()));

        let mut details = HashMap::new();
        for (tc, processed, success_count) in successful {
            let mut tc_details = HashMap::new();
            tc_details.insert("processed".to_string(), serde_json::Value::Number((*processed).into()));
            tc_details.insert("successful".to_string(), serde_json::Value::Number((*success_count).into()));
            details.insert(tc.clone(), serde_json::Value::Object(tc_details.into_iter().collect()));
        }
        statistics.insert("details".to_string(), serde_json::Value::Object(details.into_iter().collect()));

        let completion_data = serde_json::json!({
            "month": month_str,
            "timestamp": chrono::Utc::now().to_rfc3339(),
            "statistics": statistics,
            "step": "calculation_scraping",
            "method": "rust_optimized_jsonl",
            "output_format": "consolidated_jsonl"
        });

        let local_file = format!("{}/calculation_scraping_completion_{}.json", self.local_temp_dir, month_str);
        let json_content = serde_json::to_string_pretty(&completion_data)?;
        fs::write(&local_file, &json_content).await?;

        let s3_key = format!("results/{}/calculation_scraping_completion.json", month_str);
        self.s3_client
            .put_object()
            .bucket(&self.s3_bucket)
            .key(&s3_key)
            .body(json_content.into_bytes().into())
            .send()
            .await?;

        fs::remove_file(&local_file).await?;
        info!("Uploaded completion marker for {}", month_str);
        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    tracing_subscriber::fmt::init();
    let args = Args::parse();

    info!("Starting Rust calculation scraper (JSONL output) for {}", args.month);

    let scraper = PlayerCalculationScraper::new(args.s3_bucket, args.aws_region).await?;
    scraper.scrape_calculations_for_month(&args.month).await?;

    info!("Calculation scraping completed successfully for {}", args.month);
    Ok(())
}